### PR TITLE
Tdr-1130 session expire dialog

### PR DESCRIPTION
**Pull request by thanhz**
Screenshot from 2022-06-16 17-46-21

### CODE REVIEW

**Comment by thanhz** [comment]
Should the message be changed to `"You have been inactive for more than 55 Minutes"` since the session expiry is currently set to 1 hour?

**Comment by thanhz** [comment]
As per comment on the [jira card](https://national-archives.atlassian.net/browse/TDR-1108) the message has been changed to 55 minutes to reflect the 1 hour session timeout.

**Comment by thanhz** [comment]
As per comment on the [jira card](https://national-archives.atlassian.net/browse/TDR-1108) the message has been changed to 55 minutes to reflect the 1 hour session timeout.

File: npm/src/index.ts [hunk 1]
```diff
@@ -38,6 +38,8 @@ export const renderModules = async () => {
   const fileChecksContainer: HTMLDivElement | null = document.querySelector(
     ".file-check-progress"
   )
+  const timeoutDialog: HTMLDialogElement | null =
```
Comment (TomJKing) [inline npm/src/index.ts:41]: The `HTMLDialogElement` is showing its deprecated. Having a quick look, it appears this is a bug with Typescript: https://github.com/microsoft/TypeScript/issues/48267 Could you double check that this is the case.

File: npm/src/index.ts [hunk 2]
```diff
@@ -106,5 +108,49 @@ export const renderModules = async () => {
     } else {
       errorHandlingModule.handleUploadError(frontEndInfo)
     }
+  } else if (timeoutDialog) {
+    const authModule = await import("./auth")
```
Comment (TomJKing) [inline npm/src/index.ts:112]: There is some complicated logic in this function. Would it be possible to abstract this out into the existing `auth/index.ts` file. Or maybe a new file `àuth/sesssiontimeout.ts`? Can then add some more unit testing around the functionality

File: app/views/main.scala.html [hunk 3]
```diff
@@ -73,6 +73,21 @@
         </main>
     </div>
 
+    @if(isLoggedIn) {
+    <dialog class="timeout-dialog">
```
Comment (TomJKing) [inline app/views/main.scala.html:77]: At the moment on the `/upload` page the token is continuously updated, for 12 hours, to ensure a user does not get logged out whilst their files are being uploaded. There is a question as to whether:

1. this behaviour should be retained; or
2. it should be removed, as they would then get this session timeout warning.

I think this question should be asked to the wider team, including Kirsten and Pauline.

File: npm/test/index.test.ts [hunk 4]
```diff
@@ -91,3 +91,28 @@ test("renderModules does not call authorisation when the front end info is missi
 
   keycloakInstance.mockRestore()
 })
+
```
Comment (TomJKing) [inline npm/test/index.test.ts:94]: Probably need some more extensive unit tests. See suggestion about, about abstracting the functionality into separate functions to make unit testing easier.

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some comments around unit testing, and other stuff.

File: app/views/main.scala.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (thanhz) [reply app/views/main.scala.html:77]: I have tested the screenreader for the timeout message. When the model appears the screenreader will read out the message and the options.

File: npm/src/index.ts (hunk 1 above)
Comment (thanhz) [reply npm/src/index.ts:41]: The HTMLDialogElement isn't showing as deprecated for me. It does say experimental, but from what I have gathered many browsers do support this. I have also tested it out with various browsers myself.

File: app/views/main.scala.html (hunk 3 above)
Comment (thanhz) [reply app/views/main.scala.html:77]: As discussed today. We will be retaining the behavior of the upload page.

File: npm/src/index.ts (hunk 2 above)
Comment (thanhz) [reply npm/src/index.ts]: As suggested, I have moved the session timeout logic to its own file. Still working on the tests though

File: npm/test/index.test.ts (hunk 4 above)
Comment (thanhz) [reply npm/test/index.test.ts:94]: Tests finally working

File: npm/test/session-timeout.test.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  const mockUpdateToken = jest
+    .fn()
+    .mockImplementation((_: number) => new Promise((res, _) => res(true)))
+  const isTokenExpired = false
+  const now = new Date()
+  const refreshTokenParsed: KeycloakTokenParsed = {
+    exp: Math.round(now.getTime() / 1000) + 300
+  }
+
+  const mockKeycloak: Keycloak.KeycloakInstance = createMockKeycloak(
```
Comment (TomJKing) [inline npm/test/session-timeout.test.ts:71]: This is repeated code. Could it be moved out of the functions?

File: npm/test/session-timeout.test.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  const mockUpdateToken = jest
+    .fn()
+    .mockImplementation((_: number) => new Promise((res, _) => res(true)))
+  const isTokenExpired = false
+  const now = new Date()
+  const refreshTokenParsed: KeycloakTokenParsed = {
+    exp: Math.round(now.getTime() / 1000) - 300
+  }
+
+  const mockKeycloak: Keycloak.KeycloakInstance = createMockKeycloak(
```
Comment (TomJKing) [inline npm/test/session-timeout.test.ts:103]: This is repeated code. Could it be moved out of the functions?

**Review by TomJKing — COMMENTED** [review]
Looks good. Thanks for getting those jest tests in place. Just one very minor comment.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks

**Pull request merged**